catalog/spark/spark.bom

```diff
@@ -162,6 +175,87 @@ brooklyn.catalog:
           id: spark-worker
           brooklyn.config:
             spark.master.url: $brooklyn:entity("spark-master").attributeWhenReady("spark.url")
+            spark.master.rest.url: $brooklyn:entity("spark-master").attributeWhenReady("spark.rest.url")
+            spark.master.cluster.url: $brooklyn:entity("spark-master").attributeWhenReady("spark.cluster.url")
+
+  - id: spark-application
+    description: An Apache Spark Application
+    item:
+      type: org.apache.brooklyn.entity.software.base.VanillaSoftwareProcess
+      name: Spark Application
+      id: spark-application
+      brooklyn.parameters:
+        - name: spark.app.class
+          label: "Spark App Class"
+          description: |
+            Spark App Class
+          type: string
+          default: "org.apache.spark.examples.SparkPi"
+        - name: spark.app.jar
+          label: "Spark App Jar"
+          description: |
+            Spark App Jar
+          type: string
+          default: "http://dl.bintray.com/typesafe/maven-releases/org/apache/spark/spark-examples_2.11/1.6.0-typesafe-001/spark-examples_2.11-1.6.0-typesafe-001.jar"
+
+      brooklyn.initializers:
+        - type: org.apache.brooklyn.core.sensor.ssh.SshCommandSensor
+          brooklyn.config:
+            name: spark.app.submission.id
+            period: 1s
+            command: |
+              if [ -f ${RUN_DIR}/submissionId ]
+              then
+                  cat ${RUN_DIR}/submissionId
+              fi
+
+        - type: org.apache.brooklyn.core.sensor.http.HttpRequestSensor
+          brooklyn.config:
+            name: spark.app.status
+            period: 1s
+            jsonPath: driverState
+            uri:
+              '$brooklyn:formatString':
+                - '%s/v1/submissions/status/%s'
+                - $brooklyn:config("spark.master.rest.url")
+                - $brooklyn:attributeWhenReady("spark.app.submission.id")
+
+      brooklyn.config:
+
+        shell.env:
+          SPARK_CLUSTER_URL: $brooklyn:config("spark.master.cluster.url")
+          PARENT_RUN_DIR: $brooklyn:component("parent", "ignored").attributeWhenReady("run.dir")
+          SPARK_APP_CLASS: $brooklyn:config("spark.app.class")
+          SPARK_APP_JAR: $brooklyn:config("spark.app.jar")
+          RUN_DIR: $brooklyn:attributeWhenReady("run.dir")
+
+        launch.command: |
+          set +e
+          cd ${PARENT_RUN_DIR}/spark-*
+          ./bin/spark-submit --deploy-mode cluster --master "${SPARK_CLUSTER_URL}" --class "${SPARK_APP_CLASS}" "${SPARK_APP_JAR}" 2> result.txt
+
+          if [ $? -ne 0 ]; then
+            echo "An error occurred when running spark-submit, please check the parameters:"
+            echo "Class: ${SPARK_APP_CLASS}"
+            echo "Jar: ${SPARK_APP_JAR}"
+            exit 1
+          else
+            cat result.txt |grep submissionId | awk '{print $3}' | tr -d '",' > ${RUN_DIR}/submissionId
```

Comment:
So in the launch command rather than changing directory and running `./bin/spark-submit` run it from the application's own directory: `${PARENT_RUN_DIR}/spark-*/bin/spark-submit`. Then you don't have the problem of overwriting files.